pkg/ciliumidentity: Operator Managing CID Controller

ovidiutirla opened the pull request:
Add the controller for handling the Operator managing CIDs. The Cilium Identity (CID) controller running in cilium-operator and is responsible only for managing CID API objects.
* Pod events are added to Pod work queue
* Namespace events are processed immediately and added to Pod work queue
* CID events are added to CID work queue
* Processing Pod work queue items are adding items to CID work queue
* Processed CID work queue items result in mutations to CID API objects

Related CFP #27752
Reconciler PR: #33383

```release-note
Adds the operator logic for managing CIDs.
```

ovidiutirla committed:
pkg/ciliumidentity: Add CID OP Pod listener

Add the pod event listener and handler to enqueue pod reconciliation

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/ciliumidentity: Add CID OP Namespace listener

Add the namespace event listener and handler to reconcile namespace

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/ciliumidentity: Initialize labels filter

Initialize the labels filter to limit identity-relevant labels

Signed-off-by: Ovidiu Tirla <otirla@google.com>

bowei commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
-func (c *Controller) Stop(hookContext cell.HookContext) error {
+func (c *Controller) startEventProcessing() error {
+	if err := c.wp.Submit("proc-cid-events", c.processCiliumIdentityEvents); err != nil {
+		return fmt.Errorf("failed to init worker pool for CiliumIdentityEvents")
```
Do you want to chain the original error so it will be logged?

`fmt.Errorf("...: %w", err)`

If you don't have formatting done, recommend using `errors.New()` instead.

ovidiutirla replied:
chained the original error too.

bowei commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
+		return fmt.Errorf("failed to init worker pool for CiliumIdentityEvents")
+	}
+	if err := c.wp.Submit("proc-pod-events", c.processPodEvents); err != nil {
+		return fmt.Errorf("failed to init worker pool for PodEvents")
```
See comment above.

ovidiutirla replied:
chained the original error too.

bowei commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
+		return fmt.Errorf("failed to init worker pool for PodEvents")
+	}
+	if err := c.wp.Submit("proc-ces-events", c.processCiliumEndpointSliceEvents); err != nil {
+		return fmt.Errorf("failed to init worker pool for CiliumEndpointSliceEvents")
```
See comment above.

ovidiutirla replied:
chained the original error too.

bowei commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
+		return fmt.Errorf("failed to init worker pool for CiliumEndpointSliceEvents")
+	}
+	if err := c.wp.Submit("proc-ns-events", c.processNamespaceEvents); err != nil {
+		return err
```
See comment above.

ovidiutirla replied:
chained the original error too.

bowei commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
+
+func (c *Controller) startWorkQueues() error {
+	if err := c.wp.Submit("op-managing-cid-wq", c.runCIDWorker); err != nil {
+		return fmt.Errorf("failed to init worker pool for CID Worker")
```
See comment above.

ovidiutirla replied:
chained the original error too.

bowei commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
+		return fmt.Errorf("failed to init worker pool for CID Worker")
+	}
+	if err := c.wp.Submit("op-managing-pod-wq", c.runPodWorker); err != nil {
+		return fmt.Errorf("failed to init worker pool for Pod Worker")
```
See comment above.

ovidiutirla replied:
chained the original error too.

bowei commented on operator/pkg/ciliumidentity/controller_test.go:
```diff
@@ -0,0 +1,185 @@
+
+			verified, err := createPodAndVerifyCIDIsCreated(ctx, &fakeClient, cidResource)
+			if err != nil {
+				t.Errorf("unexpected error when creating pod and verifying CIDs: %v", err)
```
It might be more consistent to do something like this:

`t.Errorf("createPodAndVerifyIsCreated() = %v, want nil", err)`

to avoid drift between the human description and the func being called.

ovidiutirla replied:
updated the error message with your suggestion.

bowei commented on operator/pkg/ciliumidentity/controller_test.go:
```diff
@@ -0,0 +1,185 @@
+				t.Errorf("expected CID Is Created to be %v, but got %v", tc.enableOperatorManageCIDs, verified)
+			}
+
+			verified, err = verifyCIDUsageInCES(ctx, &fakeClient, cidResource, cesResource)
```
This is a useful idiom to potentially make it less verbose:

```
verified, err = verifyCIDUsageInCES(...)
if gotErr := err != nil; gotErr != tc.expectCESError {
  t.Errorf("verifyCIDUsageInCES() = %v; gotErr = %t, want %t", err, gotErr, tc.expectCESError)
}
```

ovidiutirla replied:
updated with your suggestion, thanks

pippolo84 replied:
The above pattern is nice, but I suggest reconsidering using it in a test. I think it would be better to keep the test code as straightforward as possible, aiming for better readability and glanceability, even at the cost of being more verbose and/or repetitive.
Furthermore, we have just two test cases here: the first one is not expecting an error while the second one is. Why not separating the two tests? We can still rely on a helper to setup common parts like the test hive.

bowei commented on operator/pkg/ciliumidentity/pod.go:
```diff
@@ -3,8 +3,114 @@
+}
+
+func (c *Controller) enqueuePodReconciliation(podKey resource.Key, delay time.Duration) {
+	if len(podKey.String()) == 0 {
```
Is this an expected condition? When will this happen? I recommend putting a comment unless this is extremely obvious why the key would be zero-length.

ovidiutirla replied:
removed. not expected to happen

bowei commented on operator/pkg/ciliumidentity/pod.go:
```diff
@@ -3,8 +3,114 @@
+	}
+
+	// Drop the pod from queue, exceeded max retries
+	c.logger.Error("Dropping the Pod from queue, exceeded maxRetries",
```
I usually like adding to the log the maxRetries for completeness for the user.

ovidiutirla replied:
added

bowei commented on operator/pkg/ciliumidentity/pod.go:
```diff
@@ -3,8 +3,114 @@
+	c.logger.Error("Failed to process Pod", logfields.Error, err)
+
+	if c.podQueue.NumRequeues(item) < maxProcessRetries {
+		c.podQueue.AddRateLimited(item)
```
Should we log something here, either a Debug or Error log? Is this expected enough that we don't need to emit anything?

I also recommend adding the number of retries that have been attempted to help debugging.

ovidiutirla replied:
appended to existing logs the number of retries and also the number of max retries.

bowei commented on operator/pkg/ciliumidentity/pod.go:
```diff
@@ -3,8 +3,114 @@
+// marks them done. You may run as many of these in parallel as you wish; the
+// work queue guarantees that they will not end up processing the same Pod at the
+// same time.
+func (c *Controller) runPodWorker(_ context.Context) error {
```
We don't use the context that is passed in. What does this context correspond to?

ovidiutirla replied:
Required by the WorkerPool, https://github.com/cilium/cilium/blob/db782e47de3e52d95d77cf4c077dfe03a31dbf2a/vendor/github.com/cilium/workerpool/workerpool.go#L93

bowei commented on operator/pkg/ciliumidentity/pod.go:
```diff
@@ -3,8 +3,114 @@
+
+	c.podQueue = workqueue.NewRateLimitingQueueWithConfig(
+		workqueue.NewItemExponentialFailureRateLimiter(defaultSyncBackOff, maxSyncBackOff),
+		workqueue.RateLimitingQueueConfig{Name: "pod"})
```
is `pod` the right name or something like "ciliumidentity_pod"

ovidiutirla replied:
updated the name

bowei commented on operator/pkg/ciliumidentity/pod.go:
```diff
@@ -3,8 +3,114 @@
+
+func (c *Controller) initPodQueue() {
+	c.logger.Info("CID controller work queue configuration for Pod",
+		logfields.WorkQueueSyncBackOff, defaultSyncBackOff)
```
If you are going to log the configuration, you should log all of the values:

* defaultSyncBackOff
* maxSyncBackOff

ovidiutirla replied:
added logs for both

bowei commented on operator/pkg/ciliumidentity/namespace.go:
```diff
@@ -3,8 +3,61 @@
+	}
+}
+
+func getNamespaceLabels(ns *slimcorev1.Namespace) labels.Labels {
```
I recommend a short unit test for this (and other helper functions)

ovidiutirla replied:
added unit tests for `getNamespaceLabels`. Are there any methods that you consider might need a unit test?

bowei commented on operator/pkg/ciliumidentity/namespace.go:
```diff
@@ -3,8 +3,61 @@
+	c.oldNSSecurityLabels[ns.Name] = newIdLabels
+
+	if err := c.reconciler.reconcileNamespace(nsResourceKey(ns.Name)); err != nil {
+		c.logger.Error("Failed to process namespaces changes", logfields.Error, err)
```
[minor] suggest making the log message closer to the code to make it easier to maintain: 

c.logger.Error("reconcileNamespace() failed", "Name", ns.Name, logfields.Error, err)

ovidiutirla replied:
I feel like changing the log message to just `reconcileNamespace()` would make logging different from the existing pattern. Added ns.Name to the logging fields though.

bowei commented on operator/pkg/ciliumidentity/namespace.go:
```diff
@@ -3,8 +3,61 @@
+	for event := range c.namespace.Events(ctx) {
+		switch event.Kind {
+		case resource.Upsert:
+			c.logger.Debug("Got Upsert Namespace event",
```
[minor] merge lines

ovidiutirla replied:
merged the lines.

bowei commented on operator/pkg/ciliumidentity/identity.go:
```diff
@@ -3,8 +3,112 @@
+}
+
+func (c *Controller) enqueueCIDReconciliation(cidKey resource.Key, delay time.Duration) {
+	if len(cidKey.String()) == 0 {
```
see similar comment in the other reconciller, why would the key be invalid?

ovidiutirla replied:
indeed, removed this check.

bowei reviewed (commented):
Added some comments -- note, I might be commenting on some practices that may be obvious to someone more familiar with the code.

ovidiutirla committed:
pkg/logging: Add WorkQueueMaxSyncBackOff to logfields

Signed-off-by: Ovidiu Tirla <otirla@google.com>

derailed commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -20,6 +22,17 @@ import (
+	// maxRetries is the number of times a work queue sync will be retried before
+	// it is dropped out of the queue.
+	maxProcessRetries = 15
+	workerPoolSize    = 6
```
How did we assess the queue size and backoff settings?

ovidiutirla replied:
`workerPoolSize` is based on the number of workers we used below,
`defaultSyncBackOff`/`maxSyncBackOff`/`maxProcessRetries` could be fine tuned further if we have any concerns.

derailed commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
+	//----------------------------V------------------V
+	// Namespace event -> Pod work queue -> CID work queue -> Mutate CID API objects
+	c.wp = workerpool.New(workerPoolSize)
+	err = c.startEventProcessing()
```
nit: inline the err check here and below

ovidiutirla replied:
added inline check

derailed commented on operator/pkg/ciliumidentity/identity.go:
```diff
@@ -3,8 +3,109 @@
+// work queue guarantees that they will not end up processing the same CID
+// at the same time
+func (c *Controller) runCIDWorker(_ context.Context) error {
+	c.logger.Info("Starting CID worker in CID controller")
```
I think we should additionally watch the passed in context  vs the controller context as this would indicate processing on the queue is being terminated?

ovidiutirla replied:
added a case for `case <-context.Done():`

derailed commented on operator/pkg/ciliumidentity/namespace.go:
```diff
@@ -3,8 +3,60 @@
+func (c *Controller) processNamespaceEvents(ctx context.Context) error {
+	for event := range c.namespace.Events(ctx) {
+		switch event.Kind {
+		case resource.Upsert:
```
do we need to do anything if a ns is deleted?

ovidiutirla replied:
If the ns is deleted, the pods will also be deleted so we do not need to reconcile the pods. No action needed on ns deletion.

derailed commented on operator/pkg/ciliumidentity/identity.go:
```diff
@@ -3,8 +3,109 @@
+		case resource.Delete:
+			c.logger.Debug("Got Delete CID event",
+				logfields.CIDName, event.Key.String())
+			c.onCiliumIdentityDeleteEvent(event.Object)
```
We trigger reconciliation with no distinction for either upsert/delete thus why do we need two separate methods? vs something like reconcile(ResourceKey, delay). 
It might be worth evaluating the design choices here so we can dry up this code and streamline the needs to reimplement similar logic for namespace, pod and cids, ... 
Perhaps leveraging generics or tidying the interfaces could help us dry this up a little more.

ovidiutirla replied:
Indeed, we do not need two separate methods, moved both under the same method. We still need to have different queues for CIDs and Pods as we need to take different actions. For pods we end up with `reconcilePod` while for CIDs we end up with `reconcileCID`.  

We could potentially refactor the existing implementation of the queues and have a single queue and rework the existing worker logic to differentiate based on the resource key, but I'm not sure how much value it would bring or it would just make the code less readable.

ovidiutirla replied:
Indeed, I gave I try to your suggestion and code looks much better, I extracted it to a separate interface. I think we might just need to spawn two workers now to archive the same effect. But for now let's go with only one and we can easily adjust later to have a configurable number of workers.

derailed commented on operator/pkg/ciliumidentity/identity.go:
```diff
@@ -3,8 +3,109 @@
+
+func (c *Controller) processCiliumIdentityEvents(ctx context.Context) error {
+	for event := range c.ciliumIdentity.Events(ctx) {
+		switch event.Kind {
```
Is there a possibility here that we either get a nil event or event.Object is nil?

ovidiutirla replied:
We shouldn't get nil events as the events are emitted from resource and are of type Event[T]. And also a event.Object should not be nil. 
Similar logic is used in multiple places (like indetitygc, clustermesh, ces,...) without any guard for nil.

derailed reviewed (commented):
@ovidiutirla Nice work! 
I could be out of context here but I think the design could use some TLC to dry up the code and eliminate the need to reimplement similar logic for the various watched types. Either leveraging generics or tidying up the interfaces.

Perhaps something like??

```go
type QueuedResource[T K8sResource] interface {
   Key() ResourceKey
   ShouldReconcile(EventType, T) bool
   ...
}
```

ovidiutirla committed:
pkg/ciliumidentity: Add CID OP CES listener

Add the CES event listener and handler to reconcile CID

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/ciliumidentity: Add CID OP CID listener

Add the CID event listener and handler to reconcile CID

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/ciliumidentity: Add CID OP controller

Add the controller for handling the Operator managing CIDs.
The Cilium Identity (CID) controller running in cilium-operator and is responsible only for managing CID API objects.
* Pod events are added to Pod work queue
* Namespace events are processed immediately and added to Pod work queue
* CID events are added to CID work queue
* Processing Pod work queue items are adding items to CID work queue
* Processed CID work queue items result in mutations to CID API objects

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/ciliumidentity: Add CID OP controller QueuedItem

Add a generic type interface for queuing Pods or CIDs that need to
reconciled

Signed-off-by: Ovidiu Tirla <otirla@google.com>

pippolo84 commented on operator/pkg/ciliumidentity/pod.go:
```diff
@@ -16,3 +21,14 @@ type PodItem struct {
+func (c *Controller) processPodEvents(ctx context.Context) error {
+	for event := range c.pod.Events(ctx) {
+		if event.Kind == resource.Upsert || event.Kind == resource.Delete {
+			c.logger.Debug("Got Pod event", logfields.Type, event.Kind, logfields.K8sPodName, event.Key.String())
```
We should use `c.logger.With(logfields.Type, event.Kind, logfields.K8sPodName, event.Key.String()).Debug("Got Pod event")` to separate log message from the parameters.

Same for all the usages of the logger.

ovidiutirla replied:
Thanks, replaced all with the suggestion.

ovidiutirla replied:
Just seen now, if we use logger.With() will return a new logger (as it clones the existing one to add these new fields). Not sure how expensive is to clone the logger in go just a FYI. 

Looking at https://github.com/cilium/cilium/pull/33121 it seems like they did not use the With() for log lines that contain parameters that are only used once. 

Do you feel like is the best way to move forward in this case? Where we do not have multiple log lines with the same parameters?

pippolo84 replied:
Yeah, you're right, didn't notice that. Although the perf impact should be minimal (especially because most of the logging for operator identity mgmt is at debug level) better be consistent across the codebase. :+1:

pippolo84 commented on operator/pkg/ciliumidentity/ciliumendpointslice.go:
```diff
@@ -0,0 +1,73 @@
+				logfields.CESName, event.Key.String())
+			c.onCiliumEndpointSliceDelete(event.Object)
+		}
+		event.Done(nil)
```
Can be simplified with:

```go
		var idsWithNoCESUsage []int64

		switch event.Kind {
		case resource.Upsert:
			c.logger.Debug("Got Upsert CES event",
				logfields.CESName, event.Key.String())
			idsWithNoCESUsage = c.reconciler.cidUsageInCES.ProcessCESUpsert(event.Object.Name, event.Object.Endpoints)
		case resource.Delete:
			c.logger.Debug("Got Delete CES event",
				logfields.CESName, event.Key.String())
			idsWithNoCESUsage = c.reconciler.cidUsageInCES.ProcessCESDelete(event.Object.Name, event.Object.Endpoints)
		}

		for _, cid := range idsWithNoCESUsage {
			cidName := strconv.Itoa(int(cid))

			c.logger.Info("Reconciling CID as it is no longer used in CESs", logfields.CIDName, cidName)
			c.enqueueReconciliation(CIDItem{cidResourceKey(cidName)}, 0)
		}

		event.Done(nil)
```

This way we can also get rid of the two functions below

ovidiutirla replied:
suggestion applied.

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -10,14 +10,28 @@ import (
+)
+
+const (
+	// defaultSyncBackOff is the default backoff period for cesSync calls.
```
The comment is mentioning CES, I guess that's a leftover from a copy-paste

ovidiutirla replied:
thanks, fixed the comment

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -10,14 +10,28 @@ import (
+const (
+	// defaultSyncBackOff is the default backoff period for cesSync calls.
+	defaultSyncBackOff = 1 * time.Second
+	// maxSyncBackOff is the max backoff period for cesSync calls.
```
Same here

ovidiutirla replied:
thanks, fixed the comment

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -53,6 +67,13 @@ type Controller struct {
+	// processed multiple times concurrently, and if a resource key is added
+	// multiple times before it can be processed, this will only be processed once.
+	resourceQueue workqueue.RateLimitingInterface
+	wp            *workerpool.WorkerPool
```
I think we should prefer 5 distinct `OneShot` hive/job instead of a workerpool. Hive jobs have several advantages (like integration with hive start and stop hooks, metrics reporting, pprof labels and so on). Take a look at `pkg/datapath/iptables` or `operator/pkg/lbipam` for some examples or look at https://github.com/cilium/hive for the complete docs.

ovidiutirla replied:
thanks, replaced with job.oneshot

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -110,12 +133,139 @@ func (c *Controller) Start(_ cell.HookContext) error {
+	return true
+}
+
+func (c *Controller) handleErr(err error, item QueuedItem) {
```
I don't think there's much value in moving the error management in a separate function, as this is called only from `processNextItem`. Probably better to move this code back into the caller.

ovidiutirla replied:
It increases the readability by reducing long methods. I would keep this as a separate method. I've added the check if err is nil in the parent method.

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -110,12 +133,139 @@ func (c *Controller) Start(_ cell.HookContext) error {
+	}
+
+	retries := c.resourceQueue.NumRequeues(item)
+	c.logger.Error("Failed to process resource item",
```
Maybe a warning is more appropriate since we are gonna retry. Let's leave the error for when we go above the retry threshold.

ovidiutirla replied:
converted to a warn.

pippolo84 reviewed (commented):
Thanks @ovidiutirla ! :100: 
Made a first pass and dropped some comments, gonna do another one after the first feedback is addresses. :+1:

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +134,137 @@ func (c *Controller) Start(_ cell.HookContext) error {
+		c.handleErr(err, qItem)
+	}
+
+	c.resourceQueue.Forget(item)
```
Shouldn't we forget the item only when `retries >= maxProcessRetries` ?

ovidiutirla replied:
indeed, I think make sense to remove the handleErr method though as you previously mentioned, made the logs inline to reduce the code pollution.

pippolo84 commented on operator/pkg/ciliumidentity/controller_test.go:
```diff
@@ -0,0 +1,189 @@
+				cell.Invoke(func(
+					c *k8sClient.FakeClientset,
+					_ resource.Resource[*slim_corev1.Namespace],
+					podResource resource.Resource[*slim_corev1.Pod],
```
Those are unused, they can be removed

ovidiutirla replied:
nice spot, removed, thanks.

pippolo84 commented on operator/pkg/ciliumidentity/controller_test.go:
```diff
@@ -0,0 +1,189 @@
+	cidStore, _ := cidResource.Store(ctx)
+	err := testutils.WaitUntil(func() bool {
+		return len(cidStore.List()) > 0
+	}, time.Second)
```
Although one sec should be enough for most cases, I suggest to raise this to at least 5 secs. This can help avoiding possible flakes while running the tests in the CI, where the performance characteristic is unpredictable.
Moreover, this is just a threshold, the check is still carried out once every 10 msecs in `WaitUntil`, so it won't make the test slower.

Same for all the occurrences of `testutils.WaitUntil`

ovidiutirla replied:
extracted `WaitUntilTimeout` as const and increased the timeout to 5 seconds. good to know, we'll keep this in mind for future.

pippolo84 commented on operator/pkg/ciliumidentity/controller_test.go:
```diff
@@ -0,0 +1,189 @@
+				t.Errorf("verifyCIDUsageInCES() = %v; gotErr = %t, want %t", err, gotErr, tc.expectCESError)
+			}
+
+			if verified != tc.enableOperatorManageCIDs {
```
Breaking this test into two separate ones (see https://github.com/cilium/cilium/pull/33933#discussion_r1691834203) should also have the additional benefit of letting `verifyCIDUsageInCES` return only an error. Same for `createPodAndVerifyCIDIsCreated`

ovidiutirla replied:
reworked the tests with your suggestions

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -10,19 +10,34 @@ import (
 type queueOperations interface {
-	enqueueCIDReconciliation(cidKey resource.Key, delay time.Duration)
-	enqueuePodReconciliation(podKey resource.Key, delay time.Duration)
+	enqueueReconciliation(item QueuedItem, delay time.Duration)
```
Nit: there is just one op: `enqueueReconciliation`. We should rename the interface to `queueOperation` or similar.

ovidiutirla replied:
renamed.

pippolo84 commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -108,15 +134,137 @@ func (c *Controller) Start(_ cell.HookContext) error {
+
+	if err != nil {
+		c.handleErr(err, qItem)
+	}
```
If we extend the `QueuedItem` interface with:

```go
  Reconcile(reconciler *reconciler) error
```

and add those methods to `PodItem` and `CIDItem`, respectively:

```go
func (p PodItem) Reconcile(reconciler *reconciler) error {
       return reconciler.reconcilePod(p.key)
}

func (c CIDItem) Reconcile(reconciler *reconciler) error {
       return reconciler.reconcileCID(c.key)
}
```

then the snippet above can be simplified with:

```go
        if err := qItem.Reconcile(c.reconciler); err != nil {
                c.handleErr(err, qItem)
        }
```

Not a big difference, but I find this a bit cleaner. WDYT?

ovidiutirla replied:
Sounds good, updated the implementation.

pippolo84 reviewed (commented):
Thanks for the follow up. Left some comments inline, but nothing major. :+1: 

One thing to consider is the current design of the reconciler, separate from the controller. Despite doing this makes the reconciler easier to unit test, I don't like the manual management of its lifecycle.
We are currently forced to manage its context despite its lifecycle is strictly tied with the one of the controller, managed by the hive.
Additionally, having this separation between the controller and the reconciler forces us to add the queueOperations interface. This probably suggests that separating them may not be the best choice.
Anyway, let's leave this for now, we can still reevaluate later.

Another question: I appreciate that the code is nicely unit tested, but being a pretty complex change is there any plan to add some e2e testing (maybe in cilium-cli) in a follow up PR?

pippolo84 reviewed (approved):
LGTM, thanks! :100: 

You should also add the relevant release-note label and text. See point 11 and 12 of https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submit-pr

pchaigno reviewed (approved):
Approving for the two lines of cilium/cli (in logfields).
